File: Template_Skill/src/main/scala/com/neong/voice/wolfpack/CalendarConversation.scala
package com.neong.voice.wolfpack

import collection.JavaConverters._

import com.neong.voice.model.base.Conversation

import com.wolfpack.database.DbConnection

class CalendarConversation extends Conversation {
  import com.amazon.speech.speechlet.{
    IntentRequest,
    Session,
    SpeechletResponse
  }

  import CalendarConversation._
  import ConversationIntent._

  for (intent <- ConversationIntent.values) {
    supportedIntentNames.add(intent.toString)
  }

  override def respondToIntentRequest(
    intentReq: IntentRequest,
    session: Session
  ): SpeechletResponse = {
    val intent = intentReq.getIntent
    val conversationIntent = ConversationIntent.withName(intent.getName)

    val db = new DbConnection("DbCredentials.xml")

    if (db.getRemoteConnection) {
      val responseSsml = conversationIntent match {
        case NextEventIntent => CalendarDataProvider.nextEventSsml(db)
      }
      Conversation.newTellResponse(responseSsml.toString, true)
    } else {
      Conversation.newTellResponse("Sorry, I'm on break right now.", false)
    }
  }
}

object CalendarConversation {
  import java.sql.Timestamp
  import java.time.ZoneId
  import java.time.format.DateTimeFormatter

  object ConversationIntent extends Enumeration {
    type ConversationIntent = Value
    val NextEventIntent = Value("NextEventIntent")
  }

  private object CalendarDataProvider {
    import scala.xml._

    def nextEventSsml(db: DbConnection, limit: Int = 5): Elem = {
      val result = db.runQuery(
        "SELECT count(*), summary, start, name " +
          "FROM event_info " +
          "WHERE start = (" +
            "SELECT start FROM event_info WHERE start > now() LIMIT 1" +
          ") " +
          "GROUP BY summary, start, name " +
          s"LIMIT $limit;"
      )

      Option(result) match {
        case Some(events) => eventsSsml(events, limit)
        case None => noEventsSsml
      }
    }

    private final val LocalZone = ZoneId.of("America/Los_Angeles")
    private final val TimeFormatter = DateTimeFormatter.ofPattern("h:mm a")
    private final val DateFormatter = DateTimeFormatter.ofPattern("????MMdd")
    private final val DayFormatter = DateTimeFormatter.ofPattern("EEEE")

    private type ResultsMap = java.util.Map[String, java.util.Vector[Object]]

    private final val noEventsSsml =
      <speak>Sorry, there's nothing coming up on the calendar.</speak>

    private def eventsSsml(events: ResultsMap, limit: Int): Elem = {
      val count = events.get("count").get(0).asInstanceOf[Long].toInt
      <speak>
        OK, the next event{if (count > 1) "s are" else " is"}
        {for (i <- 0 until count) yield eventSsml(events, i)}
        {if (count > limit)
          <s>There are {count - limit} more events starting at this time.</s>
        }
      </speak>
    }

    private def eventSsml(events: ResultsMap, index: Int): Elem = {
      val what = events.get("summary").get(index).asInstanceOf[String]
      val when = events.get("start").get(index).asInstanceOf[Timestamp]
      val location = events.get("name").get(index).asInstanceOf[String]

      val dateTime = when.toInstant.atZone(LocalZone)
      val day = dateTime.format(DayFormatter)
      val date = dateTime.format(DateFormatter)
      val time = dateTime.format(TimeFormatter)

      val where = Option(location) match {
        case Some(place) => place
        case None => "Sonoma State University"
      }

      <s>
        {what}
        on {day} <say-as interpret-as="date">{date}</say-as>
        at <say-as interpret-as="time">{time}</say-as>
        at {where}
      </s>
    }
  }
}

File: Template_Skill/src/main/clojure/com/neong/voice/wolfpack/CalendarConversation.clj
(ns com.neong.voice.wolfpack.CalendarConversation
  (:gen-class :extends com.neong.voice.model.base.Conversation
              :post-init register-intents)
  (:import [java.sql Timestamp]
           [java.time ZoneId]
           [java.time.format DateTimeFormatter]
           [com.amazon.speech.slu Intent]
           [com.amazon.speech.speechlet IntentRequest Session SpeechletResponse]
           [com.neong.voice.model.base Conversation]
           [com.neong.voice.wolfpack CalendarConversation]
           [com.wolfpack.database DbConnection]))


(defn -register-intents
  [^CalendarConversation this]
  (.add (.supportedIntentNames this) "NextEventIntent"))


;;
;; Date/time formatting helpers
;;

(def zone-id (ZoneId/of "America/Los_Angeles"))
(def day-formatter (DateTimeFormatter/ofPattern "EEEE"))
(def date-formatter (DateTimeFormatter/ofPattern "????MMdd"))
(def time-formatter (DateTimeFormatter/ofPattern "h:mm a"))

(defn at-zone [when]
  (.atZone (.toInstant when) zone-id))

(defn format-day [when]
  (.format when day-formatter))

(defn format-date [when]
  (.format when date-formatter))

(defn format-time [when]
  (.format when time-formatter))


;;
;; SSML helpers
;;

(defn speak [content]
  (str "<speak>" content "</speak>"))

(defn say-as [interpret-as content]
  (str "<say-as interpret-as=\"" interpret-as "\">" content "</say-as>"))

;;(defn section [content]
;;  (str "<s>" content "</s>"))

(defn format-calendar-response
  [what when where]
  (let [zoned (at-zone when)
        day (format-day zoned)
        date (format-date zoned)
        time (format-time zoned)]
    (speak
     (str "Okay, the next event is " what
          " on " day " " (say-as "date" date)
          " at " (say-as "time" time)
          ", at " where "."))))


;;
;; Database helpers
;;

(def next-event-query "SELECT * FROM event_info WHERE start > now() LIMIT 1;")

(defn select-next-event [db]
  (if-let [result (.runQuery db next-event-query)]
    (let [what (.get result "summary")
          when (or (.get result "start") "Sonoma State University")
          where (.get result "name")]
      {:what  (cast String    (.get what  0))
       :when  (cast Timestamp (.get when  0))
       :where (cast String    (.get where 0))})))


;;
;; Response helper
;;

(def abort-response (Conversation/newTellResponse "Sorry, I'm on break." false))

(defn lookup-response [db]
  (if-let [{what :what, when :when, where :where} (select-next-event db)]
    (Conversation/newTellResponse (format-calendar-response what when where) true)))

;;
;; Request handler
;;

(defn -respondToIntentRequest
  [^CalendarConversation this ^IntentRequest request ^Session session]
  (or
   (let [db (DbConnection. "DbConnection.xml")]
     (if (.getRemoteConnection db)
       (lookup-response db)))
   abort-response))
